### ISSUE TITLE
Cannot download deb package via apt: 'NoneType' object has no attribute 'read'

### ISSUE THREAD

**Issue by Frederick888** [ansible/ansible#23800]
##### ISSUE TYPE
 - Bug Report

##### COMPONENT NAME
`apt`

##### ANSIBLE VERSION
```
ansible 2.3.0.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = Default w/o overrides
  python version = 2.7.13 (default, Dec 21 2016, 07:16:46) [GCC 6.2.1 20160830]
```

##### CONFIGURATION
```
couchdb_deb: couchdb_2.0.0-8.deb
```

##### OS / ENVIRONMENT
Arch Linux (Client) / Ubuntu 16.04 (Server)

##### SUMMARY
Cannot download deb and install via apt module

##### STEPS TO REPRODUCE
Run the following task

```yaml
- name: Install CouchDB
  become: yes 
  become_user: root
  apt:
    deb: https://onee3.org/deb/{{ couchdb_deb }}
```

##### EXPECTED RESULTS
Ansible could install deb after downloading it.

##### ACTUAL RESULTS
It failed to download.

```
<x.x.x.x> (0, '\r\n{"msg": "Failure downloading https://onee3.org/deb/couchdb_2.0.0-8.deb, \'NoneType\' object has no attribute \'read\'", "failed": true, "invocation": {"module_args": {"dpkg_options": "force-confdef,force-confold", "upgrade": null, "force": false, "package": null, "purge": false, "allow_unauthenticated": false, "state": "present", "autoremove": null, "update_cache": null, "deb": "https://onee3.org/deb/couchdb_2.0.0-8.deb", "only_upgrade": false, "cache_valid_time": 0, "default_release": null, "install_recommends": null}}}\r\n', 'OpenSSH_7.4p1, OpenSSL 1.0.2k  26 Jan 2017\r\ndebug1: Reading configuration data /home/frederick/.ssh/config\r\ndebug1: Reading configuration data /etc/ssh/ssh_config\r\ndebug1: auto-mux: Trying existing master\r\ndebug2: fd 3 setting O_NONBLOCK\r\ndebug2: mux_client_hello_exchange: master version 4\r\ndebug3: mux_client_forwards: request forwardings: 0 local, 0 remote\r\ndebug3: mux_client_request_session: entering\r\ndebug3: mux_client_request_alive: entering\r\ndebug3: mux_client_request_alive: done pid = 13582\r\ndebug3: mux_client_request_session: session request sent\r\ndebug1: mux_client_request_session: master session id: 2\r\ndebug3: mux_client_read_packet: read header failed: Broken pipe\r\ndebug2: Received exit status from master 0\r\nShared connection to x.x.x.x closed.\r\n')
fatal: [avocet.couchdb.cloud]: FAILED! => {
    "changed": false, 
    "failed": true, 
    "invocation": {
        "module_args": {
            "allow_unauthenticated": false, 
            "autoremove": null, 
            "cache_valid_time": 0, 
            "deb": "https://onee3.org/deb/couchdb_2.0.0-8.deb", 
            "default_release": null, 
            "dpkg_options": "force-confdef,force-confold", 
            "force": false, 
            "install_recommends": null, 
            "only_upgrade": false, 
            "package": null, 
            "purge": false, 
            "state": "present", 
            "update_cache": null, 
            "upgrade": null
        }
    }, 
    "msg": "Failure downloading https://onee3.org/deb/couchdb_2.0.0-8.deb, 'NoneType' object has no attribute 'read'"
}
```

**Comment by Akasurde** [comment]
@Frederick888 Can you please try with #23821 ?

**Comment by Frederick888** [comment]
@Akasurde I applied the patch and got 403 instead:

```
<x.x.x.x> (0, b'\r\n{"msg": "Failed to download https://onee3.org/deb/couchdb_2.0.0-8.deb, HTTP Error 403: Forbidden", "failed": true, "invocation": {"module_args": {"dpkg_options": "force-confdef,force-confold", "upgrade": null, "force": false, "package": null, "purge": false, "allow_unauthenticated": false, "state": "present", "autoremove": null, "update_cache": null, "deb": "https://onee3.org/deb/couchdb_2.0.0-8.deb", "only_upgrade": false, "cache_valid_time": 0, "default_release": null, "install_recommends": null}}}\r\n', b'OpenSSH_7.4p1, OpenSSL 1.0.2k  26 Jan 2017\r\ndebug1: Reading configuration data /home/frederick/.ssh/config\r\ndebug1: Reading configuration data /etc/ssh/ssh_config\r\ndebug1: auto-mux: Trying existing master\r\ndebug2: fd 3 setting O_NONBLOCK\r\ndebug2: mux_client_hello_exchange: master version 4\r\ndebug3: mux_client_forwards: request forwardings: 0 local, 0 remote\r\ndebug3: mux_client_request_session: entering\r\ndebug3: mux_client_request_alive: entering\r\ndebug3: mux_client_request_alive: done pid = 3637\r\ndebug3: mux_client_request_session: session request sent\r\ndebug1: mux_client_request_session: master session id: 2\r\ndebug3: mux_client_read_packet: read header failed: Broken pipe\r\ndebug2: Received exit status from master 0\r\nShared connection to x.x.x.x closed.\r\n')                    
fatal: [avocet.couchdb.cloud]: FAILED! => {
    "changed": false,
    "failed": true,
    "invocation": {
        "module_args": {
            "allow_unauthenticated": false,
            "autoremove": null,
            "cache_valid_time": 0,
            "deb": "https://onee3.org/deb/couchdb_2.0.0-8.deb",
            "default_release": null,
            "dpkg_options": "force-confdef,force-confold",
            "force": false,
            "install_recommends": null,
            "only_upgrade": false,
            "package": null,
            "purge": false,
            "state": "present",
            "update_cache": null,
            "upgrade": null
        }
    },
    "msg": "Failed to download https://onee3.org/deb/couchdb_2.0.0-8.deb, HTTP Error 403: Forbidden"
}
```
Should I checkout the whole repo?

**Comment by Akasurde** [comment]
@Frederick888 HTTP response code 403, means this patch is working. There is problem with URL, try with some other URL/Server.

**Comment by Frederick888** [comment]
But what is the problem of the URL? I can download it from the server using cURL and wget

--
Best Regards,
Frederick Zhang

Email: frederick888@tsundere.moe

**Comment by Akasurde** [comment]
@Frederick888 paste output of ansible-playbook with `-vvvv`

**Comment by Frederick888** [comment]
@Akasurde Full log here. Btw, the URL works with `get_url` module
[log.txt](https://github.com/ansible/ansible/files/946125/log.txt)

**Comment by Akasurde** [comment]
@Frederick888 There is some problem with URL itself. 
I am able download and install package using following playbook using patch. 

```
---
- name: Apt deb download
  hosts: myself
  tasks:
    - name: Download and install hello
      apt:
        deb: http://ftp.us.debian.org/debian/pool/main/h/hello/hello_2.10-1+b1_amd64.deb
```

**Comment by Frederick888** [comment]
@Akasurde So what *is* the problem?

```
$ curl -I https://onee3.org/deb/couchdb_2.0.0-8.deb
HTTP/1.1 200 OK
Date: Fri, 21 Apr 2017 08:49:56 GMT
Content-Type: application/octet-stream
Content-Length: 17969492
Connection: keep-alive
Set-Cookie: __cfduid=dbb3130f4ddb4fc3714269f7be3c448a91492764595; expires=Sat, 21-Apr-18 08:49:55 GMT; path=/; domain=.onee3.org; HttpOnly
Last-Modified: Tue, 18 Apr 2017 11:39:40 GMT
ETag: "58f5fafc-1123154"
Strict-Transport-Security: max-age=63072000; includeSubdomains; preload
Accept-Ranges: bytes
Server: cloudflare-nginx
CF-RAY: 352f17c1bf2c1d86-MEL

$ curl -I http://ftp.us.debian.org/debian/pool/main/h/hello/hello_2.10-1+b1_amd64.deb
HTTP/1.1 200 OK
Date: Fri, 21 Apr 2017 08:51:28 GMT
Server: Apache/2.2.22 (Ubuntu)
Last-Modified: Tue, 14 Mar 2017 23:43:34 GMT
ETag: "3b20741-da58-54ab96563d580"
Accept-Ranges: bytes
Content-Length: 55896
Content-Type: application/x-debian-package
X-Pad: avoid browser bug
```
I can't see any significant differences between them.

And actually, I temporarily disabled CloudFlare but I didn't see any requests coming in from Ansible. The log is working normally as I could see my requests via cURL and wget:

```
x.x.x.x - y.y.y.y - [21/Apr/2017:08:56:51 +0000] "HEAD /deb/couchdb_2.0.0-8.deb HTTP/1.1" 200 0 "-" "curl/7.53.1"
x.x.x.x - y.y.y.y - [21/Apr/2017:08:57:41 +0000] "GET /deb/couchdb_2.0.0-8.deb HTTP/1.1" 200 17969492 "-" "Wget/1.19.1 (linux-gnu)"
```

**Comment by aleskiontherun** [comment]
@Akasurde @sivel Same problem with Ansible 2.3.0/2.3.1. Worked fine with 2.2.0. Throws 403 with latest devel.

```
- name: Install
  apt:
    deb: https://downloads.wkhtmltopdf.org/0.12/0.12.2.1/wkhtmltox-0.12.2.1_linux-trusty-amd64.deb
```
Tried quoting the URL, still 403.

**Comment by Frederick888** [comment]
Tbh it was hilarious for this issue to be closed by a completely unrelated commit without actually knowing the reason of it.

**Comment by Akasurde** [comment]
@dizeee Could you please elaborate more on your problem? if problem persists, you can re-open the issue.

**Comment by aleskiontherun** [comment]
@Akasurde Yes, the problem persists. Running Ansible on macOS Sierra 10.12.5.
```
python version = 2.7.10 (default, Feb  7 2017, 00:08:15) [GCC 4.2.1 Compatible Apple LLVM 8.0.0 (clang-800.0.34)]
```

The task:
```
- name: Install
  apt:
    deb: https://downloads.wkhtmltopdf.org/0.12/0.12.2.1/wkhtmltox-0.12.2.1_linux-trusty-amd64.deb
```

Here's output with Ansible 2.3.1.0:
```
TASK [wkhtmltopdf : Install] ****************************************************************************************************************
task path: /***/provisioner/roles/wkhtmltopdf/tasks/main.yml:7
Using module file /***/ansible/ansible/lib/ansible/modules/packaging/os/apt.py
<x.x.x.x> ESTABLISH SSH CONNECTION FOR USER: None
<x.x.x.x> SSH: EXEC ssh -vvv -F ./ssh.conf -o ControlMaster=auto -o ControlPersist=30m -o StrictHostKeyChecking=no -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o ConnectTimeout=20 -o 'ControlPath=~/.ssh/connection-***-ansible-%r@%h:%p' x.x.x.x '/bin/sh -c '"'"'sudo sudo -H -S -n -u root /bin/sh -c '"'"'"'"'"'"'"'"'echo BECOME-SUCCESS-fuwogdxhufkagrqimyvyeertqhkslkua; /usr/bin/python'"'"'"'"'"'"'"'"' && sleep 0'"'"''
<x.x.x.x> (1, '\n{"msg": "Failure downloading https://downloads.wkhtmltopdf.org/0.12/0.12.2.1/wkhtmltox-0.12.2.1_linux-trusty-amd64.deb, \'NoneType\' object has no attribute \'read\'", "failed": true, "invocation": {"module_args": {"dpkg_options": "force-confdef,force-confold", "upgrade": null, "force": false, "package": null, "purge": false, "allow_unauthenticated": false, "state": "present", "autoremove": null, "update_cache": null, "deb": "https://downloads.wkhtmltopdf.org/0.12/0.12.2.1/wkhtmltox-0.12.2.1_linux-trusty-amd64.deb", "only_upgrade": false, "cache_valid_time": 0, "default_release": null, "install_recommends": null}}}\n', 'OpenSSH_7.4p1, LibreSSL 2.5.0\r\ndebug1: Reading configuration data ./ssh.conf\r\ndebug1: ./ssh.conf line 1: Applying options for x.x.x.*\r\ndebug1: auto-mux: Trying existing master\r\ndebug2: fd 3 setting O_NONBLOCK\r\ndebug2: mux_client_hello_exchange: master version 4\r\ndebug3: mux_client_forwards: request forwardings: 0 local, 0 remote\r\ndebug3: mux_client_request_session: entering\r\ndebug3: mux_client_request_alive: entering\r\ndebug3: mux_client_request_alive: done pid = 35871\r\ndebug3: mux_client_request_session: session request sent\r\ndebug1: mux_client_request_session: master session id: 2\r\ndebug3: mux_client_read_packet: read header failed: Broken pipe\r\ndebug2: Received exit status from master 1\r\n')
fatal: [x.x.x.x]: FAILED! => {
    "changed": false,
    "failed": true,
    "invocation": {
        "module_args": {
            "allow_unauthenticated": false,
            "autoremove": null,
            "cache_valid_time": 0,
            "deb": "https://downloads.wkhtmltopdf.org/0.12/0.12.2.1/wkhtmltox-0.12.2.1_linux-trusty-amd64.deb",
            "default_release": null,
            "dpkg_options": "force-confdef,force-confold",
            "force": false,
            "install_recommends": null,
            "only_upgrade": false,
            "package": null,
            "purge": false,
            "state": "present",
            "update_cache": null,
            "upgrade": null
        }
    }
}

MSG:

Failure downloading https://downloads.wkhtmltopdf.org/0.12/0.12.2.1/wkhtmltox-0.12.2.1_linux-trusty-amd64.deb, 'NoneType' object has no attribute 'read'
```

And with latest devel branch:
```
TASK [wkhtmltopdf : Install] ****************************************************************************************************************
task path: /***/provisioner/roles/wkhtmltopdf/tasks/main.yml:7
Using module file /***/ansible/ansible/lib/ansible/modules/packaging/os/apt.py
<x.x.x.x> ESTABLISH SSH CONNECTION FOR USER: None
<x.x.x.x> SSH: EXEC ssh -vvv -F ./ssh.conf -o ControlMaster=auto -o ControlPersist=30m -o StrictHostKeyChecking=no -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o ConnectTimeout=20 -o 'ControlPath=~/.ssh/connection-***-ansible-%r@%h:%p' x.x.x.x '/bin/sh -c '"'"'sudo sudo -H -S -n -u root /bin/sh -c '"'"'"'"'"'"'"'"'echo BECOME-SUCCESS-ujruezrxyszodncqvbguxciqefxuzrut; /usr/bin/python'"'"'"'"'"'"'"'"' && sleep 0'"'"''
<x.x.x.x> (1, '\n{"msg": "Failed to download https://downloads.wkhtmltopdf.org/0.12/0.12.2.1/wkhtmltox-0.12.2.1_linux-trusty-amd64.deb, HTTP Error 403: Forbidden", "failed": true, "changed": false, "invocation": {"module_args": {"dpkg_options": "force-confdef,force-confold", "autoremove": false, "force": false, "package": null, "autoclean": false, "purge": false, "allow_unauthenticated": false, "state": "present", "upgrade": null, "update_cache": null, "deb": "https://downloads.wkhtmltopdf.org/0.12/0.12.2.1/wkhtmltox-0.12.2.1_linux-trusty-amd64.deb", "only_upgrade": false, "cache_valid_time": 0, "default_release": null, "install_recommends": null}}}\n', 'OpenSSH_7.4p1, LibreSSL 2.5.0\r\ndebug1: Reading configuration data ./ssh.conf\r\ndebug1: ./ssh.conf line 1: Applying options for x.x.x.*\r\ndebug1: auto-mux: Trying existing master\r\ndebug2: fd 3 setting O_NONBLOCK\r\ndebug2: mux_client_hello_exchange: master version 4\r\ndebug3: mux_client_forwards: request forwardings: 0 local, 0 remote\r\ndebug3: mux_client_request_session: entering\r\ndebug3: mux_client_request_alive: entering\r\ndebug3: mux_client_request_alive: done pid = 35871\r\ndebug3: mux_client_request_session: session request sent\r\ndebug1: mux_client_request_session: master session id: 2\r\ndebug3: mux_client_read_packet: read header failed: Broken pipe\r\ndebug2: Received exit status from master 1\r\n')
fatal: [x.x.x.x]: FAILED! => {
    "changed": false,
    "failed": true,
    "invocation": {
        "module_args": {
            "allow_unauthenticated": false,
            "autoclean": false,
            "autoremove": false,
            "cache_valid_time": 0,
            "deb": "https://downloads.wkhtmltopdf.org/0.12/0.12.2.1/wkhtmltox-0.12.2.1_linux-trusty-amd64.deb",
            "default_release": null,
            "dpkg_options": "force-confdef,force-confold",
            "force": false,
            "install_recommends": null,
            "only_upgrade": false,
            "package": null,
            "purge": false,
            "state": "present",
            "update_cache": null,
            "upgrade": null
        }
    }
}

MSG:

Failed to download https://downloads.wkhtmltopdf.org/0.12/0.12.2.1/wkhtmltox-0.12.2.1_linux-trusty-amd64.deb, HTTP Error 403: Forbidden
```

I don't have rights to reopen the issue.

**Comment by Akasurde** [comment]
@dizeee This fix is not backported to Ansible 2.3 branch, that is why it is not working there.

Could you please open a new issue for '403 forbidden' issue?

**Comment by aleskiontherun** [comment]
@Akasurde #26239
